docs/TerraformTaintPlugin.md

```diff
@@ -0,0 +1,46 @@
+## [TerraformTaintPlugin](../src/TerraformTaintPlugin.groovy)
+
+Enable this plugin to add `TAINT_RESOURCE` and `UNTAINT_RESOURCE` parameters
+to the build. If a resource path is provided via one of those parameters, then
+the `terraform plan` command will be preceded by the corresponding Terraform
+command to taint or untaint the appropriate resources.
+
+Note that the untaint command will take precedence, so if for some reason the
+same resource is placed in both parameters, it will be tainted and immediately
+untainted, resulting in no change.
+
+There are several ways to customize where and when the taint/untaint can run:
+
+* `onlyOnOriginRepo()`: This method takes in a string of the form
+  `<userOrOrg>/<repoName>`. When specified, the plugin will only apply when
+  the current checkout is from the same repository and not a fork or other
+  clone.
```

Comment:
I don't fully understand the notion of `onlyOnOriginRepo`.  If this option did not exist, how would one taint the resources from a fork? (Meaning, I'm not clear how one could taint a resource from a fork, so I'm not clear what problem this option is addressing?)

Reply:
The main reason this is in is because this was a requirement for the work I'm doing for this feature. Really, we don't want someone forking the project to work on it and having their builds affect state. I suppose the `onlyOnOriginRepo` could be extracted out into a separate plugin that could apply to all things.

Reply:
"Really, we don't want someone forking the project to work on it and having their builds affect state." <- Is this notion different from the notion of, "We don't want someone to run `apply` from their fork"?  

If so, the solution for that is to restrict "apply" to the master branch.  That's conceptually consistent with other plugins that affect state, and negates the need for an additional "originRepo" concept.  

I'm less concerned about extracting it to another plugin, and more concerned about introducing a concept that partially overlaps with the notion of branches.

Reply:
Just wanted to weigh in a bit here...

branch != fork. I think the main concern here is for folks like me who set up their personal user as a GitHub Organization Folder in some Jenkins... and then when I fork a repo, it starts building under my Org as well as where it "should" build. So, while not being able to taint on branches other than master is probably good, it doesn't really solve the fork problem. It's really only a "problem" if someone provides parameters to the wrong job, but that's disturbingly easy to do, especially since the PR status links sent to GitHub will be links back to the fork in Jenkins.

All that being said, this might actually be a larger issue that effects pretty much everything other than "plan"... which is that you can do things on the master branch of a fork that you shouldn't be able to do anywhere other than the master branch of the upstream/canonical/origin repository...

Reply:
"this might actually be a larger issue that effects pretty much everything other than "plan"... which is that you can do things on the master branch of a fork that you shouldn't be able to do anywhere other than the master branch of the upstream/canonical/origin repository..."  <-- I agree.  If that's indeed a problem, taint/untaint doesn't seem like the appropriate place to start addressing the problem.

Reply:
I'll reprhase the above in a much more friendly way (rereading that, it sounds very unfriendly, which was not the intention).

If doing things on a fork/branch are a problem, it's indeed a big and fundamental problem.  I haven't been particularly concerned about it, because I feel like there are workarounds outside of this library that handle those concerns (ie: you can disable untrusted branches/forks at the Jenkins configuration level).  But maybe there are other concerns that you're thinking about, that haven't crossed my mind, and I'd be interested to hear them.

It seems like the problem and the likely concerns would be much broader than the narrow feature of taint/untaint, and likely deserves a separate issue, that can look at the problem across all existing plugins.

Reply:
Agreed that's likely a larger issue. I'll try to flesh out #356 a bit.